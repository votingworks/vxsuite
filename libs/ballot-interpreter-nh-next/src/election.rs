use std::fmt::Display;

use serde::{Deserialize, Serialize};

use crate::{ballot_card::BallotSide, types::idtype};

// import idtype macro from types.rs

idtype!(ContestId);
idtype!(OptionId);
idtype!(BallotStyleId);
idtype!(PrecinctId);

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Election {
    pub title: String,
    pub grid_layouts: Vec<GridLayout>,
    pub mark_thresholds: Option<MarkThresholds>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GridLayout {
    pub precinct_id: PrecinctId,
    pub ballot_style_id: BallotStyleId,
    pub columns: u32,
    pub rows: u32,
    pub grid_positions: Vec<GridPosition>,
}

/// A position on the ballot grid defined by timing marks and the contest/option
/// for which a mark at this position is a vote for.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", tag = "type")]
pub enum GridPosition {
    /// A pre-defined labeled option on the ballot.
    #[serde(rename_all = "camelCase", rename = "option")]
    Option {
        side: BallotSide,
        column: u32,
        row: u32,
        contest_id: ContestId,
        option_id: OptionId,
    },

    /// A write-in option on the ballot.
    #[serde(rename_all = "camelCase", rename = "write-in")]
    WriteIn {
        side: BallotSide,
        column: u32,
        row: u32,
        contest_id: ContestId,
        write_in_index: u32,
    },
}

impl Display for GridPosition {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Option { option_id, .. } => write!(f, "{option_id}"),

            Self::WriteIn { write_in_index, .. } => {
                write!(f, "Write-In {write_in_index}")
            }
        }
    }
}

impl GridPosition {
    pub const fn location(&self) -> GridLocation {
        match self {
            Self::Option {
                side, column, row, ..
            }
            | Self::WriteIn {
                side, column, row, ..
            } => GridLocation::new(*side, *column, *row),
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize)]
pub struct GridLocation {
    pub side: BallotSide,
    pub column: u32,
    pub row: u32,
}

impl GridLocation {
    pub const fn new(side: BallotSide, column: u32, row: u32) -> Self {
        Self { side, column, row }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MarkThresholds {
    pub definite: f32,
    pub marginal: f32,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_grid_location() {
        let location = GridLocation::new(BallotSide::Front, 1, 2);
        assert_eq!(location.side, BallotSide::Front);
        assert_eq!(location.column, 1);
        assert_eq!(location.row, 2);
    }

    #[test]
    fn test_grid_position() {
        let position = GridPosition::Option {
            side: BallotSide::Front,
            column: 1,
            row: 2,
            contest_id: ContestId::from("contest-1".to_string()),
            option_id: OptionId::from("option-1".to_string()),
        };
        assert_eq!(position.location().side, BallotSide::Front);
        assert_eq!(position.location().column, 1);
        assert_eq!(position.location().row, 2);
    }

    #[test]
    fn test_grid_position_option_serialization() {
        let json = r#"{
            "type": "option",
            "side": "front",
            "column": 1,
            "row": 2,
            "contestId": "contest-1",
            "optionId": "option-1"
        }"#;
        match serde_json::from_str(json).unwrap() {
            GridPosition::Option {
                side,
                column,
                row,
                contest_id,
                option_id,
            } => {
                assert_eq!(side, BallotSide::Front);
                assert_eq!(column, 1);
                assert_eq!(row, 2);
                assert_eq!(contest_id, ContestId::from("contest-1".to_string()));
                assert_eq!(option_id, OptionId::from("option-1".to_string()));
            }
            _ => panic!("expected Option"),
        }
    }

    #[test]
    fn test_grid_position_write_in_serialization() {
        let json = r#"{
            "type": "write-in",
            "side": "front",
            "column": 1,
            "row": 2,
            "contestId": "contest-1",
            "writeInIndex": 3
        }"#;
        match serde_json::from_str(json).unwrap() {
            GridPosition::WriteIn {
                side,
                column,
                row,
                contest_id,
                write_in_index,
            } => {
                assert_eq!(side, BallotSide::Front);
                assert_eq!(column, 1);
                assert_eq!(row, 2);
                assert_eq!(contest_id, ContestId::from("contest-1".to_string()));
                assert_eq!(write_in_index, 3);
            }
            _ => panic!("expected WriteIn"),
        }
    }
}
