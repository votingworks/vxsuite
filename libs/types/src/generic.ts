import check8601 from '@antongolub/iso8601';
import { z } from 'zod';
import { err, ok, Result, wrapException } from './result';

export interface Dictionary<T> {
  [key: string]: Optional<T>;
}
export type Optional<T> = T | undefined;
export interface Provider<T> {
  get(): Promise<T>;
}

/**
 * Parse `value` using `parser`. Note that this takes an object that is already
 * supposed to be of type `T`, not a JSON string. For that, use `safeParseJson`.
 *
 * @returns `Ok` when the parse succeeded, `Err` otherwise.
 */
export function safeParse<T>(
  parser: z.ZodType<T>,
  value: unknown
): Result<T, z.ZodError> {
  const result = parser.safeParse(value);

  if (!result.success) {
    return err(result.error);
  }

  return ok(result.data);
}

/**
 * Parse JSON without throwing an exception if it's malformed. On success the
 * result will be `Ok<unknown>` and you'll need to validate the result yourself.
 * Given malformed JSON, the result will be `Err<SyntaxError>`. Add a parser
 * argument to automatically validate the resulting value after deserializing
 * the JSON.
 */
export function safeParseJson(text: string): Result<unknown, SyntaxError>;
/**
 * Parse JSON and then validate the result with `parser`.
 */
export function safeParseJson<T>(
  text: string,
  parser: z.ZodType<T>
): Result<T, z.ZodError | SyntaxError>;
export function safeParseJson<T>(
  text: string,
  parser?: z.ZodType<T>
): Result<T | unknown, z.ZodError | SyntaxError> {
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch (error) {
    return wrapException(error);
  }

  return parser ? safeParse(parser, parsed) : ok(parsed);
}

/**
 * Parse `value` using `parser` or throw an exception if parsing fails.
 */
export function unsafeParse<T>(parser: z.ZodType<T>, value: unknown): T {
  return safeParse(parser, value).unsafeUnwrap();
}

/**
 * Parser `value` using `parser` or return `undefined` if parsing fails.
 */
export function maybeParse<T>(
  parser: z.ZodType<T>,
  value: unknown
): T | undefined {
  return safeParse(parser, value).ok();
}

declare const Unique: unique symbol;

/**
 * Builds a nominal type based on `Base` with tag `Tag`. This allows creating a
 * type that acts as `Base` but is not assignable from `Base`.
 *
 * @example
 *
 *   type EmailAddress = NewType<string, 'EmailAddress'>;
 *
 *   function parseEmailAddress(emailAddress: string): Optional<EmailAddress> {
 *     if (EMAIL_ADDRESS_PATTERN.test(emailAddress)) {
 *       return emailAddress as EmailAddress;
 *     }
 *     return undefined;
 *   }
 *
 *   declare function sendEmail(to: EmailAddress, subject: string, body: string): Promise<void>;
 *
 *   await sendEmail('admin@example.com', 'Hello', 'World');
 *   //              ^^^^^^^^^^^^^^^^^^^
 *   // Argument of type 'string' is not assignable to parameter of type 'EmailAddress'.
 *
 *   const to = parseEmailAddress('admin@example.com');
 *   assert(to);
 *   await sendEmail(to, 'Hello', 'World');
 */
export type NewType<Base, Tag> = Base & { readonly [Unique]: Tag };

/**
 * Creates a type that is either `T` or a `Promise` wrapping `T`.
 */
export type PromiseOr<T> = T | Promise<T>;

export type Id = string;
export const IdSchema: z.ZodSchema<Id> = z
  .string()
  .nonempty()
  .refine((id) => !id.startsWith('_'), 'IDs may not start with an underscore')
  .refine(
    (id) => /^[-_a-z\d]+$/i.test(id),
    'IDs may only contain letters, numbers, dashes, and underscores'
  );
export const ElectionHash: z.ZodSchema<string> = z
  .string()
  .nonempty()
  .refine(
    (hash) => /^[0-9a-f]*$/i.test(hash),
    'Election hashes must be hex strings containing only 0-9 and a-f'
  );
export const MachineId = z
  .string()
  .nonempty()
  .refine(
    (id) => /^[-A-Z\d]+$/.test(id),
    'Machine IDs may only contain numbers, uppercase letters, and dashes'
  );

export const Iso8601Date = z
  .string()
  .refine(check8601, 'dates must be in ISO8601 format');
export type Iso8601Timestamp = string;
export const Iso8601TimestampSchema = Iso8601Date;
