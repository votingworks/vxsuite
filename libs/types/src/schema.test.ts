import {
  election as electionSample,
  electionData,
  electionWithMsEitherNeither,
} from '../test/election';
import * as t from './election';
import { safeParse, safeParseJson, unsafeParse } from './generic';

test('parseElection', () => {
  expect(() => t.parseElection({})).toThrowError();
  expect(() => t.parseElection(electionSample)).not.toThrowError();
});

test('parsing fails on an empty object', () => {
  t.safeParseElection({}).unsafeUnwrapErr();
});

test('parsing JSON.parses a string', () => {
  expect(t.safeParseElection(electionData).unsafeUnwrap()).toEqual(
    electionSample
  );
});

test('parsing invalid JSON', () => {
  expect(t.safeParseElection('{').unsafeUnwrapErr().message).toEqual(
    'Unexpected end of JSON input'
  );
});

test('parsing JSON without a schema', () => {
  expect(safeParseJson('{}').unsafeUnwrap()).toEqual({});
});

test('parsing gives specific errors for nested objects', () => {
  expect(
    t
      .safeParseElection({
        ...electionSample,
        contests: [
          ...electionSample.contests.slice(1),
          {
            ...electionSample.contests[0],
            // give title a type it shouldn't have
            title: 42,
          },
        ],
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "invalid_union",
        "unionErrors": [
          {
            "issues": [
              {
                "code": "invalid_type",
                "expected": "string",
                "received": "number",
                "path": [
                  "contests",
                  1,
                  "title"
                ],
                "message": "Expected string, received number"
              }
            ]
          },
          {
            "issues": [
              {
                "code": "invalid_type",
                "expected": "string",
                "received": "number",
                "path": [
                  "contests",
                  1,
                  "title"
                ],
                "message": "Expected string, received number"
              },
              {
                "code": "invalid_type",
                "expected": "yesno",
                "received": "candidate",
                "path": [
                  "contests",
                  1,
                  "type"
                ],
                "message": "Expected yesno, received candidate"
              },
              {
                "code": "invalid_type",
                "expected": "string",
                "received": "undefined",
                "path": [
                  "contests",
                  1,
                  "description"
                ],
                "message": "Required"
              }
            ]
          },
          {
            "issues": [
              {
                "code": "invalid_type",
                "expected": "string",
                "received": "number",
                "path": [
                  "contests",
                  1,
                  "title"
                ],
                "message": "Expected string, received number"
              },
              {
                "code": "invalid_type",
                "expected": "ms-either-neither",
                "received": "candidate",
                "path": [
                  "contests",
                  1,
                  "type"
                ],
                "message": "Expected ms-either-neither, received candidate"
              },
              {
                "code": "invalid_type",
                "expected": "string",
                "received": "undefined",
                "path": [
                  "contests",
                  1,
                  "eitherNeitherContestId"
                ],
                "message": "Required"
              },
              {
                "code": "invalid_type",
                "expected": "string",
                "received": "undefined",
                "path": [
                  "contests",
                  1,
                  "pickOneContestId"
                ],
                "message": "Required"
              },
              {
                "code": "invalid_type",
                "expected": "string",
                "received": "undefined",
                "path": [
                  "contests",
                  1,
                  "description"
                ],
                "message": "Required"
              },
              {
                "code": "invalid_type",
                "expected": "string",
                "received": "undefined",
                "path": [
                  "contests",
                  1,
                  "eitherNeitherLabel"
                ],
                "message": "Required"
              },
              {
                "code": "invalid_type",
                "expected": "string",
                "received": "undefined",
                "path": [
                  "contests",
                  1,
                  "pickOneLabel"
                ],
                "message": "Required"
              },
              {
                "code": "invalid_type",
                "expected": "object",
                "received": "undefined",
                "path": [
                  "contests",
                  1,
                  "eitherOption"
                ],
                "message": "Required"
              },
              {
                "code": "invalid_type",
                "expected": "object",
                "received": "undefined",
                "path": [
                  "contests",
                  1,
                  "neitherOption"
                ],
                "message": "Required"
              },
              {
                "code": "invalid_type",
                "expected": "object",
                "received": "undefined",
                "path": [
                  "contests",
                  1,
                  "firstOption"
                ],
                "message": "Required"
              },
              {
                "code": "invalid_type",
                "expected": "object",
                "received": "undefined",
                "path": [
                  "contests",
                  1,
                  "secondOption"
                ],
                "message": "Required"
              }
            ]
          }
        ],
        "path": [
          "contests",
          1
        ],
        "message": "Invalid input"
      }
    ]]
  `);
});

test('ensures dates are ISO 8601-formatted', () => {
  expect(
    t
      .safeParseElection({
        ...electionSample,
        date: 'not ISO',
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "custom",
        "message": "dates must be in ISO8601 format",
        "path": [
          "date"
        ]
      }
    ]]
  `);
});

test('parsing a valid election object succeeds', () => {
  const parsed = t.safeParseElection(electionSample as unknown).unsafeUnwrap();

  // This check is here to prove TS inferred that `parsed` is an `Election`.
  expect(parsed.title).toEqual(electionSample.title);

  // Check the whole thing
  expect(parsed).toEqual(electionSample);
});

test('parsing a valid election with ms-either-neither succeeds', () => {
  const parsed = t
    .safeParseElection(electionWithMsEitherNeither as unknown)
    .unsafeUnwrap();

  // This check is here to prove TS inferred that `parsed` is an `Election`.
  expect(parsed.title).toEqual(electionWithMsEitherNeither.title);

  // Check the whole thing
  expect(parsed).toEqual(electionWithMsEitherNeither);
});

test('parsing a valid election', () => {
  expect(t.safeParseElection(electionSample).unsafeUnwrap()).toEqual(
    electionSample
  );
});

test('contest IDs cannot start with an underscore', () => {
  expect(
    safeParse(t.CandidateContestSchema, {
      ...electionSample.contests[0],
      id: '_president',
    }).unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "custom",
        "message": "IDs may not start with an underscore",
        "path": [
          "id"
        ]
      }
    ]]
  `);
});

test('allows valid mark thresholds', () => {
  t.safeParseElection({
    ...electionSample,
    markThresholds: { definite: 0.2, marginal: 0.2 },
  }).unsafeUnwrap();

  t.safeParseElection({
    ...electionSample,
    markThresholds: { definite: 0.2, marginal: 0.1 },
  }).unsafeUnwrap();
});

test('disallows invalid mark thresholds', () => {
  expect(
    t
      .safeParseElection({
        ...electionSample,
        markThresholds: { definite: 0.2, marginal: 0.3 },
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "custom",
        "message": "marginal mark threshold must be less than or equal to definite mark threshold",
        "path": [
          "markThresholds"
        ]
      }
    ]]
  `);

  expect(
    t
      .safeParseElection({
        ...electionSample,
        markThresholds: { marginal: 0.3 },
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "invalid_type",
        "expected": "number",
        "received": "undefined",
        "path": [
          "markThresholds",
          "definite"
        ],
        "message": "Required"
      }
    ]]
  `);

  expect(
    t
      .safeParseElection({
        ...electionSample,
        markThresholds: { definite: 1.2, marginal: 0.3 },
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "too_big",
        "maximum": 1,
        "type": "number",
        "inclusive": true,
        "message": "Value should be less than or equal to 1",
        "path": [
          "markThresholds",
          "definite"
        ]
      }
    ]]
  `);
});

test('allows valid adjudication reasons', () => {
  t.safeParseElection({
    ...electionSample,
    adjudicationReasons: [],
  }).unsafeUnwrap();

  t.safeParseElection({
    ...electionSample,
    adjudicationReasons: [
      t.AdjudicationReason.MarginalMark,
      t.AdjudicationReason.UninterpretableBallot,
    ],
  }).unsafeUnwrap();
});

test('disallows invalid adjudication reasons', () => {
  expect(
    t
      .safeParseElection({
        ...electionSample,
        adjudicationReasons: ['abcdefg'],
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "invalid_enum_value",
        "options": [
          "UninterpretableBallot",
          "MarginalMark",
          "Overvote",
          "Undervote",
          "WriteIn",
          "UnmarkedWriteIn",
          "BlankBallot"
        ],
        "path": [
          "adjudicationReasons",
          0
        ],
        "message": "Invalid enum value. Expected 'UninterpretableBallot' | 'MarginalMark' | 'Overvote' | 'Undervote' | 'WriteIn' | 'UnmarkedWriteIn' | 'BlankBallot', received 'abcdefg'"
      }
    ]]
  `);

  expect(
    t
      .safeParseElection({
        ...electionSample,
        adjudicationReasons: 'foooo',
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "invalid_type",
        "expected": "array",
        "received": "string",
        "path": [
          "adjudicationReasons"
        ],
        "message": "Expected array, received string"
      }
    ]]
  `);
});

test('supports ballot layout paper size', () => {
  expect(
    t
      .safeParseElection({
        ...electionSample,
        ballotLayout: {
          paperSize: 'A4',
        },
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "invalid_enum_value",
        "options": [
          "letter",
          "legal",
          "custom8.5x17"
        ],
        "path": [
          "ballotLayout",
          "paperSize"
        ],
        "message": "Invalid enum value. Expected 'letter' | 'legal' | 'custom8.5x17', received 'A4'"
      }
    ]]
  `);

  expect(
    t
      .safeParseElection({
        ...electionSample,
        ballotLayout: 'letter',
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "invalid_type",
        "expected": "object",
        "received": "string",
        "path": [
          "ballotLayout"
        ],
        "message": "Expected object, received string"
      }
    ]]
  `);
});

test('parsing validates district references', () => {
  expect(
    t
      .safeParseElection({
        ...electionSample,
        districts: [{ id: 'DIS', name: 'DIS' }],
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "custom",
        "path": [
          "ballotStyles",
          0,
          "districts",
          0
        ],
        "message": "Ballot style '1' has district 'D', but no such district is defined. Districts defined: [DIS]."
      }
    ]]
  `);
});

test('parsing validates precinct references', () => {
  expect(
    t
      .safeParseElection({
        ...electionSample,
        precincts: [{ id: 'PRE', name: 'PRE' }],
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "custom",
        "path": [
          "ballotStyles",
          0,
          "precincts",
          0
        ],
        "message": "Ballot style '1' has precinct 'P', but no such precinct is defined. Precincts defined: [PRE]."
      }
    ]]
  `);
});

test('parsing validates contest party references', () => {
  const contest = electionSample.contests.find(
    ({ id }) => id === 'CC'
  ) as t.CandidateContest;
  const remainingContests = electionSample.contests.filter(
    (c) => contest !== c
  );

  expect(
    t
      .safeParseElection({
        ...electionSample,
        contests: [
          {
            ...contest,
            partyId: 'not-a-party',
          },
          ...remainingContests,
        ],
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "custom",
        "path": [
          "contests",
          0,
          "partyId"
        ],
        "message": "Contest 'CC' has party 'not-a-party', but no such party is defined. Parties defined: [PARTY]."
      }
    ]]
  `);
});

test('parsing validates candidate party references', () => {
  const contest = electionSample.contests.find(
    ({ id }) => id === 'CC'
  ) as t.CandidateContest;
  const remainingContests = electionSample.contests.filter(
    (c) => contest !== c
  );

  expect(
    t
      .safeParseElection({
        ...electionSample,
        contests: [
          {
            ...contest,
            candidates: [
              ...contest.candidates.slice(1),
              {
                ...contest.candidates[0],
                partyId: 'not-a-party',
              },
            ],
          },
          ...remainingContests,
        ],
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "custom",
        "path": [
          "contests",
          0,
          "candidates",
          0,
          "partyId"
        ],
        "message": "Candidate 'C' in contest 'CC' has party 'not-a-party', but no such party is defined. Parties defined: [PARTY]."
      }
    ]]
  `);
});

test('validates uniqueness of district ids', () => {
  expect(
    t
      .safeParseElection({
        ...electionSample,
        districts: [...electionSample.districts, ...electionSample.districts],
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "custom",
        "path": [
          "districts",
          1,
          "id"
        ],
        "message": "Duplicate district 'D' found."
      }
    ]]
  `);
});

test('validates uniqueness of ballot style ids', () => {
  expect(
    safeParse(t.BallotStylesSchema, [
      ...electionSample.ballotStyles,
      ...electionSample.ballotStyles,
    ]).unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "custom",
        "path": [
          1,
          "id"
        ],
        "message": "Duplicate ballot style '1' found."
      }
    ]]
  `);
});

test('validates uniqueness of precinct ids', () => {
  expect(
    t
      .safeParseElection({
        ...electionSample,
        precincts: [...electionSample.precincts, ...electionSample.precincts],
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "custom",
        "path": [
          "precincts",
          1,
          "id"
        ],
        "message": "Duplicate precinct 'P' found."
      }
    ]]
  `);
});

test('validates uniqueness of contest ids', () => {
  expect(
    t
      .safeParseElection({
        ...electionSample,
        contests: [...electionSample.contests, ...electionSample.contests],
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "custom",
        "path": [
          "contests",
          2,
          "id"
        ],
        "message": "Duplicate contest 'CC' found."
      },
      {
        "code": "custom",
        "path": [
          "contests",
          3,
          "id"
        ],
        "message": "Duplicate contest 'YNC' found."
      }
    ]]
  `);
});

test('validates uniqueness of party ids', () => {
  expect(
    t
      .safeParseElection({
        ...electionSample,
        parties: [...electionSample.parties, ...electionSample.parties],
      })
      .unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "custom",
        "path": [
          "parties",
          1,
          "id"
        ],
        "message": "Duplicate party 'PARTY' found."
      }
    ]]
  `);
});

test('validates uniqueness of candidate ids within a contest', () => {
  const contest = electionSample.contests[0] as t.CandidateContest;

  expect(
    safeParse(t.CandidateContestSchema, {
      ...contest,
      candidates: [...contest.candidates, ...contest.candidates],
    }).unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "custom",
        "path": [
          "candidates",
          1,
          "id"
        ],
        "message": "Duplicate candidate 'C' found."
      }
    ]]
  `);
});

test('validates admin cards have hex-encoded hashes', () => {
  unsafeParse(t.AdminCardDataSchema, {
    t: 'admin',
    h: 'd34db33f',
  });
  expect(
    safeParse(t.AdminCardDataSchema, {
      t: 'admin',
      h: 'not hex',
    }).unsafeUnwrapErr()
  ).toMatchInlineSnapshot(`
    [Error: [
      {
        "code": "custom",
        "message": "Election hashes must be hex strings containing only 0-9 and a-f",
        "path": [
          "h"
        ]
      }
    ]]
  `);
});

test('safeParseElectionDefinition computes the election hash', () => {
  expect(
    t.safeParseElectionDefinition(electionData).unsafeUnwrap().electionHash
  ).toMatchInlineSnapshot(
    `"d5366378eeccc2fd38953e6e34c3069dea0dca4b7a8f5c789f3d108dc1807d3c"`
  );
});

test('safeParseElectionDefinition error result', () => {
  expect(t.safeParseElectionDefinition('').err()).toBeDefined();
});

test('specifying write-in candidates', () => {
  const candidateContest: t.CandidateContest = {
    id: 'CC',
    type: 'candidate',
    title: 'CC',
    section: 'Section',
    districtId: unsafeParse(t.DistrictIdSchema, 'D'),
    allowWriteIns: true,
    seats: 1,
    candidates: [
      {
        id: 'C',
        name: 'C',
      },
      {
        id: '__write-in-0',
        name: 'W',
        isWriteIn: true,
      },
    ],
  };

  unsafeParse(t.CandidateContestSchema, candidateContest);
});

test('specifying all write-in candidates is required if any are specified', () => {
  const candidateContest: t.CandidateContest = {
    id: 'CC',
    type: 'candidate',
    title: 'CC',
    section: 'Section',
    districtId: unsafeParse(t.DistrictIdSchema, 'D'),
    allowWriteIns: true,
    seats: 2,
    candidates: [
      {
        id: 'C',
        name: 'C',
      },
      {
        id: '__write-in-0',
        name: 'W',
        isWriteIn: true,
      },
    ],
  };

  expect(
    safeParse(t.CandidateContestSchema, candidateContest).unsafeUnwrapErr()
      .errors[0].message
  ).toEqual(
    'Contest has 1 write-in candidate(s), but 2 seat(s) are available.'
  );
});

test('no write-in candidates may be specified if write-ins are not allowed', () => {
  const candidateContest: t.CandidateContest = {
    id: 'CC',
    type: 'candidate',
    title: 'CC',
    section: 'Section',
    districtId: unsafeParse(t.DistrictIdSchema, 'D'),
    allowWriteIns: false,
    seats: 1,
    candidates: [
      {
        id: 'C',
        name: 'C',
      },
      {
        id: '__write-in-0',
        name: 'W',
        isWriteIn: true,
      },
    ],
  };

  expect(
    safeParse(t.CandidateContestSchema, candidateContest).unsafeUnwrapErr()
      .errors[0].message
  ).toEqual(`Contest 'CC' does not allow write-ins.`);
});
