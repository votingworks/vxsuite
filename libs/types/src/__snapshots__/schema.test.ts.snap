// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contest IDs cannot start with an underscore 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "IDs may not start with an underscore",
    "path": [
      "id"
    ]
  }
]]
`;

exports[`ensures election date is YYYY-MM-DD 1`] = `
[ZodError: [
  {
    "code": "custom",
    "message": "Date must be in the format YYYY-MM-DD",
    "path": [
      "date"
    ]
  }
]]
`;

exports[`parsing gives specific errors for nested objects 1`] = `
[ZodError: [
  {
    "code": "invalid_union",
    "unionErrors": [
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "number",
            "path": [
              "contests",
              1,
              "title"
            ],
            "message": "Expected string, received number"
          }
        ],
        "name": "ZodError"
      },
      {
        "issues": [
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "number",
            "path": [
              "contests",
              1,
              "title"
            ],
            "message": "Expected string, received number"
          },
          {
            "received": "candidate",
            "code": "invalid_literal",
            "expected": "yesno",
            "path": [
              "contests",
              1,
              "type"
            ],
            "message": "Invalid literal value, expected \\"yesno\\""
          },
          {
            "code": "invalid_type",
            "expected": "string",
            "received": "undefined",
            "path": [
              "contests",
              1,
              "description"
            ],
            "message": "Required"
          },
          {
            "code": "invalid_type",
            "expected": "object",
            "received": "undefined",
            "path": [
              "contests",
              1,
              "yesOption"
            ],
            "message": "Required"
          },
          {
            "code": "invalid_type",
            "expected": "object",
            "received": "undefined",
            "path": [
              "contests",
              1,
              "noOption"
            ],
            "message": "Required"
          }
        ],
        "name": "ZodError"
      }
    ],
    "path": [
      "contests",
      1
    ],
    "message": "Invalid input"
  }
]]
`;

exports[`parsing validates candidate party references 1`] = `
[ZodError: [
  {
    "code": "custom",
    "path": [
      "contests",
      0,
      "candidates",
      0,
      "partyIds",
      0
    ],
    "message": "Candidate 'C' in contest 'CC' has party 'not-a-party', but no such party is defined. Parties defined: [PARTY]."
  }
]]
`;

exports[`parsing validates contest party references 1`] = `
[ZodError: [
  {
    "code": "custom",
    "path": [
      "contests",
      0,
      "partyId"
    ],
    "message": "Contest 'CC' has party 'not-a-party', but no such party is defined. Parties defined: [PARTY]."
  }
]]
`;

exports[`parsing validates district references 1`] = `
[ZodError: [
  {
    "code": "custom",
    "path": [
      "ballotStyles",
      0,
      "districts",
      0
    ],
    "message": "Ballot style '1' has district 'D', but no such district is defined. Districts defined: [DIS]."
  }
]]
`;

exports[`parsing validates precinct references 1`] = `
[ZodError: [
  {
    "code": "custom",
    "path": [
      "ballotStyles",
      0,
      "precincts",
      0
    ],
    "message": "Ballot style '1' has precinct 'P', but no such precinct is defined. Precincts defined: [PRE]."
  }
]]
`;

exports[`supports ballot layout paper size 1`] = `
[ZodError: [
  {
    "received": "A4",
    "code": "invalid_enum_value",
    "options": [
      "letter",
      "legal",
      "custom-8.5x17",
      "custom-8.5x18",
      "custom-8.5x21",
      "custom-8.5x22"
    ],
    "path": [
      "ballotLayout",
      "paperSize"
    ],
    "message": "Invalid enum value. Expected 'letter' | 'legal' | 'custom-8.5x17' | 'custom-8.5x18' | 'custom-8.5x21' | 'custom-8.5x22', received 'A4'"
  }
]]
`;

exports[`supports ballot layout paper size 2`] = `
[ZodError: [
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "string",
    "path": [
      "ballotLayout"
    ],
    "message": "Expected object, received string"
  }
]]
`;

exports[`validates uniqueness of ballot style ids 1`] = `
[ZodError: [
  {
    "code": "custom",
    "path": [
      1,
      "id"
    ],
    "message": "Duplicate ballot style '1' found."
  }
]]
`;

exports[`validates uniqueness of candidate ids within a contest 1`] = `
[ZodError: [
  {
    "code": "custom",
    "path": [
      "candidates",
      1,
      "id"
    ],
    "message": "Duplicate candidate 'C' found."
  }
]]
`;

exports[`validates uniqueness of contest ids 1`] = `
[ZodError: [
  {
    "code": "custom",
    "path": [
      "contests",
      2,
      "id"
    ],
    "message": "Duplicate contest 'CC' found."
  },
  {
    "code": "custom",
    "path": [
      "contests",
      3,
      "id"
    ],
    "message": "Duplicate contest 'YNC' found."
  },
  {
    "code": "custom",
    "path": [
      "contests",
      2,
      "yes/noOption",
      "id"
    ],
    "message": "Duplicate yes/no contest option 'YNC-option-yes' found."
  },
  {
    "code": "custom",
    "path": [
      "contests",
      3,
      "yes/noOption",
      "id"
    ],
    "message": "Duplicate yes/no contest option 'YNC-option-no' found."
  }
]]
`;

exports[`validates uniqueness of district ids 1`] = `
[ZodError: [
  {
    "code": "custom",
    "path": [
      "districts",
      1,
      "id"
    ],
    "message": "Duplicate district 'D' found."
  }
]]
`;

exports[`validates uniqueness of party ids 1`] = `
[ZodError: [
  {
    "code": "custom",
    "path": [
      "parties",
      1,
      "id"
    ],
    "message": "Duplicate party 'PARTY' found."
  }
]]
`;

exports[`validates uniqueness of precinct ids 1`] = `
[ZodError: [
  {
    "code": "custom",
    "path": [
      "precincts",
      1,
      "id"
    ],
    "message": "Duplicate precinct 'P' found."
  }
]]
`;
